### PR TITLE
Added Server Sync to most configuration values.  Fixed publish.ps1 to…

### PR DESCRIPTION
**Pull request by Dracbjorn**
### Publishing Fix

Added a new build type of `DebugServer` which will push the mod to both the local Valheim Client as well as the local Valheim Dedicated Server installation for testing server side features.  

There are now new Post-build events that execute the `publish.ps1` script located in the root of the project.  That script has been modified slightly to both support the new `DebugServer` build type as well as copy the assets folder to the mod directory (that was previously only being done during a release build).

The Post-build command line is as follows:

```powershell.exe -NoProfile -ExecutionPolicy Unrestricted -command '$(SolutionDir)\publish.ps1 $(ConfigurationName) $(TargetDir) $(TargetFileName) "C:\Program Files (x86)\Steam\steamapps\common\Valheim" $(ProjectDir)'```

### Server Sync

The server based configuration synchronization was implemented based of Jotunn's documentation found here:  https://valheim-modding.github.io/Jotunn/tutorials/config.html#synced-configurations

Effectively, all that had to change was the following:

Original Config Value
```
allowed = plugin.Config.Bind("Client config", "SpectralShroudAllowed",
                true, new ConfigDescription("Whether crafting a Spectral Shroud is allowed or not."));
```
Server Sync Config Value
```
allowed = plugin.Config.Bind("Server config", "SpectralShroudAllowed",
                true, new ConfigDescription("Whether crafting a Spectral Shroud is allowed or not.", null,
                new ConfigurationManagerAttributes { IsAdminOnly = true }));
```
I'm not 100% sure if it was needed but this was also added to the `Awake()` method:

```
         // You can subscribe to a global event when config got synced initially and on changes
            SynchronizationManager.OnConfigurationSynchronized += (obj, attr) =>
            {
                if (attr.InitialSynchronization)
                {
                    Jotunn.Logger.LogMessage("Initial Config sync event received");
                }
                else
                {
                    Jotunn.Logger.LogMessage("Config sync event received");
                }
            };
 ```
 
###  Testing

To test these changes, install Valheim Dedicated Server on Steam.  Make sure you install BepInEx and Jotunn into your new Valheim Server installation prior.

Once that is done, build the project using `DebugServer`.   The publish script should copy everything to both the server and the client.

Start your server via Steam.  Then start your Valheim client and connect to your local Dedicated server.

Experiment with changing the config in the Valheim Dedicated Server Installation and the client's config to see if the client config values actually override the server's config.  Here is an example server config attached below:
            
[com.chebgonaz.FriendlySkeletonWand.txt](https://github.com/jpw1991/Friendly-Skeleton-Wand/files/10367272/com.chebgonaz.FriendlySkeletonWand.txt)

Just copy the same config to each installation directory prior to launching but make the values different in the client config to see if it works.

### CODE REVIEW

File: FriendlySkeletonWand/BasePlugin.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 using System.IO;
 using System.Linq;
 using UnityEngine;
+using static Mono.Security.X509.X520;
```
Comment (jpw1991) [inline FriendlySkeletonWand/BasePlugin.cs:22]: Can you do me a favor and try removing all changes from this `BasePlugin.cs` file and test again? I suspect that all has been added here is some logging and it probably isn't needed. If everything still works, let's omit that.

**Review by jpw1991 — CHANGES_REQUESTED** [review]
Thanks for the hard work. Can you please do the following:

1. Remove the changes from `BasePlugin.cs` and test to see if it still works. I suspect all that's been added here is some logging and if that's the case, we can omit that.
2. Don't commit the changes to:
  - `FriendlySkeletonWand.csproj`
  - `app.config`
  - `packages.config`
because these are unique to your setup and will likely break mine.

Feel free to disagree on any point and we'll discuss. Thanks again

**Comment by Dracbjorn** [comment]
> 

Ok, regarding the csproj file.  You will need to add this or the new publishing fix won't work:

```<PropertyGroup>
    <PostBuildEvent>powershell.exe -NoProfile -ExecutionPolicy Unrestricted -command '$(SolutionDir)\publish.ps1 $(ConfigurationName) $(TargetDir) $(TargetFileName) "C:\Program Files (x86)\Steam\steamapps\common\Valheim" $(ProjectDir)'</PostBuildEvent>
  </PropertyGroup>```

**Comment by jpw1991** [comment]
> Ok, regarding the csproj file. You will need to add this or the new publishing fix won't work

There's an interesting thought actually. I develop on Linux so I don't use the powershell stuff. Instead I use my publish.sh script. I wonder if that property group gets ignored on Linux or whether it errors out.

**Comment by jpw1991** [comment]
> > Ok, regarding the csproj file. You will need to add this or the new publishing fix won't work
> 
> There's an interesting thought actually. I develop on Linux so I don't use the powershell stuff. Instead I use my publish.sh script. I wonder if that property group gets ignored on Linux or whether it errors out.

Turns out it does make errors :( what a shame. Is it possible to add that nice line to your personal csproj.user file?

image

**Comment by Dracbjorn** [comment]
> I found a solution for both platforms that works in the `csproj.user` file.

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'DebugServer|AnyCPU'">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <AllowUnsafeBlocks>false</AllowUnsafeBlocks>
    <Prefer32Bit>false</Prefer32Bit>
  </PropertyGroup>
  <PropertyGroup Condition="'$(OS)' == 'Windows_NT'">
    <RunPostBuildEvent>OnOutputUpdated</RunPostBuildEvent>
    <PostBuildEvent>powershell.exe -NoProfile -ExecutionPolicy Unrestricted -command '$(SolutionDir)\publish.ps1 $(ConfigurationName) $(TargetDir) $(TargetFileName) $(VALHEIM_INSTALL) $(ProjectDir)'</PostBuildEvent>
  </PropertyGroup>
  <PropertyGroup Condition="'$(OS)' != 'Windows_NT'">
    <RunPostBuildEvent>OnOutputUpdated</RunPostBuildEvent>
    <PostBuildEvent>$(SolutionDir)publish.sh $(TargetDir) $(TargetFileName) /home/joshua/.local/share/Steam/steamapps/common/Valheim $(ProjectDir) /home/joshua/.local/share/Steam/steamapps/common/Valheim/BepInEx/plugins"</PostBuildEvent>
  </PropertyGroup>
</Project>
```

**Pull request merged**